src/OrleansServiceBus/Providers/Streams/EventHub/EventHubCacheEvictionStrategy.cs

```diff
@@ -0,0 +1,157 @@
+        /// For ordered reliable queues we shouldn't need to notify on every purged event, only on the last event 
+        ///   of every set of events that get purged.
+        /// </summary>
+        public Action<CachedEventHubMessage?, CachedEventHubMessage?> OnPurged { get; set; }
+
+        /// <inheritdoc />
+        public void OnBlockAllocated(IDisposable newBlock)
+        {
+            var newBuffer = newBlock as FixedSizeBuffer;
+            this.inUseBuffers.Enqueue(newBuffer);
+            this.currentBuffer = newBuffer;
+            newBuffer.SetPurgeAction(this.OnFreeBlockRequest);
+        }
+
+        /// <inheritdoc />
+        public void PerformPurge(DateTime nowUtc, IDisposable purgeRequest)
+        {
+            //if the cache is empty, then nothing to purge, return
+            if (this.PurgeObservable.IsEmpty)
+                return;
+            int itemsPurged = 0;
+            CachedEventHubMessage neweswtMessageInCache = this.PurgeObservable.Newest.Value;
+            CachedEventHubMessage? lastMessagePurged = null;
+            while (!this.PurgeObservable.IsEmpty)
+            {
+                var oldestMessageInCache = this.PurgeObservable.Oldest.Value;
+                if (!ShouldPurge(ref oldestMessageInCache, ref neweswtMessageInCache, nowUtc))
+                {
+                    break;
+                }
+                lastMessagePurged = oldestMessageInCache;
+                itemsPurged++;
+                this.PurgeObservable.RemoveOldestMessage();
+            }
+            //if nothing got purged, return
+            if (itemsPurged == 0)
+                return;
+
+            //purge finished, time to conduct follow up actions 
+            OnPurged?.Invoke(lastMessagePurged, this.PurgeObservable.Newest);
+            UpdatePurgedBuffers(lastMessagePurged, this.PurgeObservable.Oldest, itemsPurged > 0);
+            if (this.logger.IsVerbose)
+            { 
+                var itemCountAfterPurge = this.PurgeObservable.ItemCount;
+                var itemCountBeforePurge = itemCountAfterPurge + itemsPurged;
+                ReportPurge(itemCountBeforePurge, itemCountAfterPurge);
+            }
+        }
+
+        private void UpdatePurgedBuffers(CachedEventHubMessage? lastMessagePurged, CachedEventHubMessage? oldestMessageInCache, bool itemsGotPurged)
+        {
+            //if nothing purged, then no buffer was purged
+            //if items got purged, then potencially some buffer was purged
+            if (itemsGotPurged)
+            {
+                var IdOfLastPurgedBuffer = lastMessagePurged.Value.Segment.Array;
+                // IdOfLastBufferInCache will be null if cache is empty after purge
+                var IdOfLastBufferInCache = oldestMessageInCache.HasValue ? oldestMessageInCache.Value.Segment.Array : null;
+                //all buffer older than LastPurgedBuffer should be purged 
+                while (this.inUseBuffers.Peek().Id != IdOfLastPurgedBuffer)
```

Comment:
it is not possible here for inUseBUffers to be empty, since itemsGotPurged is true here. If there's items got purged, then definitely inUseBufffers is not empty. 
But you are right, for maintenance, we should have a empty check here